### PR TITLE
impl http event in config

### LINKED ISSUE #4: Move CI to github actions

### PR DESCRIPTION
**Pull request by artur-barsegyan**
## Motivation

During of development the "Try Tarantool" service we need to give an opportunity for a user to export his first functions via HTTP and get the first result without installing `tarantool` and `tarantoolctl`.   

This PR extends the current mechanism and adds a new type of event called `http`. 

## TODO

- [x] Remove HTTP handler if the new config doesn't exist any bindings with that endpoint

Also closes #4

### CODE REVIEW

File: test/banking_config/extensions/config.yml [hunk 1]
```diff
@@ -7,12 +7,24 @@ functions:
...
     handler: account_add
     events:
     - binary: {path: account_add}
 
+  http_account_add:
+    module: extensions.banking
+    handler: http_account_add
+    events:
+    - http: {path: /account_add, method: GET}
+
```
Comment (rosik) [inline test/banking_config/extensions/config.yml]: Here too. POST instead of GET.

And please, provide an example in the readme, and write a better pull request description.

File: test/banking_config/extensions/config.yml [hunk 2]
```diff
@@ -7,12 +7,24 @@ functions:
     events:
     - binary: {path: customer_add}
 
+  http_customer_add:
+    module: extensions.banking
+    handler: http_customer_add
+    events:
+    - http: {path: /customer_add, method: GET}
```
Comment (rosik) [inline test/banking_config/extensions/config.yml]: I guess it should be POST instead of GET

File: test/banking_test.lua [hunk 3]
```diff
@@ -88,8 +89,16 @@ function g.test_banking()
 
     srv.net_box:call('transfer_money', {1, 2, 20})
 
+    http_client.get(string.format('localhost:%s/customer_add?customer_id=%s&fullname=%s',
+                                              srv.http_port, '3', 'Artur'))
+    http_client.get(string.format('localhost:%s/account_add?customer_id=%s&account_id=%s&name=%s',
+                                              srv.http_port, '3', '3', 'default'))
```
Comment (rosik) [inline test/banking_test.lua]: It could be more concise if you used json body instead of `query_param`.

There're lots of good kv home task examples, e.g. https://github.com/vanyarock01/remote-web-storage/

File: extensions.lua [hunk 4]
```diff
@@ -152,6 +153,20 @@ local function process_config(conf)
         end
 
         for _, event in ipairs(fconf.events) do
+            if event.http == nil then
+                -- do nothing
```
Comment (rosik) [inline extensions.lua]: It should be covered with a test. `cartridge.cfg({http_enabled = false})` is an important corner case that may result in a bug, see https://github.com/tarantool/migrations/issues/26.

File: extensions.lua [hunk 5]
```diff
@@ -152,6 +153,20 @@ local function process_config(conf)
...
+                    return nil, ExtensionConfigError:new(
+                        "Invalid extensions config: " ..
+                        "collision of http event '%s'" ..
+                        " to handle function '%s'",
+                        event.http.path, fname
+                    )
+            else
+                httpd:route({ path = event.http.path, method = event.http.method}, fn)
+                ret.exports[event.http.path] = fn
+            end
```
Comment (rosik) [inline extensions.lua]: What about removing the endpoint from the config? It should be implemented and tested.

File: test/common_test.lua [hunk 6]
```diff
@@ -214,6 +216,43 @@ function g.test_functions_errors()
...
+                }}
+            }}
+        })
+    }, {
+        filename = 'extensions/box.lua',
+        content = "return { cat = function() return { status = 200, body = \"Meow\" } end}"
+    }}
+
+    set_sections(config)
+    local res = http.get(("localhost:%s/cat"):format(g.cluster.main_server.http_port))
```
Comment (rosik) [inline test/common_test.lua]: Move it to a local variable

File: test/common_test.lua [hunk 7]
```diff
@@ -214,6 +216,43 @@ function g.test_functions_errors()
...
+                module = 'extensions.box',
+                handler = 'cat',
+                events = {{
+                    http = { path = '/cat', method = 'GET' }
+                }}
+            }}
+        })
+    }, {
+        filename = 'extensions/box.lua',
+        content = "return { cat = function() return { status = 200, body = \"Meow\" } end}"
```
Comment (rosik) [inline test/common_test.lua]: Single quotes (outer) would fit better here

File: test/common_http_test.lua [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+            }}
+        })
+        conf['extensions/box.lua'] = "return { cat = function() end }"
+
+        return cartridge.config_patch_clusterwide(conf)
+    ]])
+
+    t.assert_is(
+        "Cartridge was inited with the HTTP server turned off. " ..
+        "Set the option \"http_enabled\" to \"true\" value and apply config again.",
```
Comment (rosik) [inline test/common_http_test.lua]: Single quotes

File: test/common_test.lua [hunk 9]
```diff
@@ -214,6 +216,43 @@ function g.test_functions_errors()
...
+    set_sections({{
+        filename = 'extensions/config.yml',
+        content = yaml.encode({
+            functions = {}
+        })
+    }})
+    t.assert_is(
+        http.get(("localhost:%s/cat"):format(g.cluster.main_server.http_port)).status,
+        404
+    )
```
Comment (rosik) [inline test/common_test.lua]:
```
Failed tests:                                                                                                                                                                                              
-------------                                                                                                                                                                                              
                                                                                                                                                                                                           
1) common.test_http_event_removing                                                                                                                                                                         
/opt/cartridge-extensions/test/common_test.lua:252: expected and actual object should not be different                                   
Expected: 404                                                                                                                                                                                              
Received: 200                                                                                        
stack traceback:                                                                                     
        /opt/cartridge-extensions/test/common_test.lua:252: in function 'common.test_http_event_removing'               
        ...                                                                                                                                                                                                
        [C]: in function 'xpcall'                                                                                                                                                                          

```

Did you run it?

File: test/common_http_test.lua [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+
+    g.cluster:start()
+end
+
+g.after_all = function()
+    g.cluster:stop()
+    fio.rmtree(g.cluster.datadir)
+end
+
+local error_prefix = "Invalid extensions config: "
```
Comment (rosik) [inline test/common_http_test.lua]: Unused. Please pay attention to `luacheck`. There're lots of warnings across the patch.

File: extensions.lua [hunk 11]
```diff
@@ -152,6 +156,28 @@ local function process_config(conf)
         end
 
         for _, event in ipairs(fconf.events) do
+            if event.http == nil then
+                -- do nothing
+            elseif cartridge_args.http_enabled == false then
+                return nil, ExtensionConfigError:new(
+                    "Cartridge was inited with the HTTP server turned off. " ..
+                    "Set the option \"http_enabled\" to \"true\" value and apply config again."
+                )
```
Comment (rosik) [inline extensions.lua]: Does it mean if we have a single instance with HTTP disabled we can't apply config?

File: extensions.lua [hunk 12]
```diff
@@ -203,16 +247,32 @@ local function apply_config(conf)
...
+        local to_remove = {}
+        for path, _ in pairs(vars.http_exports) do
+            for i, route in ipairs(httpd.routes) do
+                if path:match(route.match) or string.match(path .. '/', route.match) then
+                    table.insert(to_remove, i)
+                end
+            end
+        end
+        remove_old_routes(httpd.routes, to_remove)
+    end
```
Comment (rosik) [inline extensions.lua]: It's ambiguous

File: extensions.lua [hunk 13]
```diff
@@ -152,6 +156,28 @@ local function process_config(conf)
...
+                    "collision of http event '%s'" ..
+                    " to handle function '%s'",
+                    event.http.path, fname
+                )
+            else
+                ret.http_exports[event.http.path] = {
+                    path = event.http.path,
+                    method = event.http.method,
+                    func = fn
+                }
```
Comment (rosik) [inline extensions.lua]: why do you duplicate `event.http.path`?

File: test/common_test.lua [hunk 14]
```diff
@@ -214,6 +216,43 @@ function g.test_functions_errors()
             }),
         }}
     )
+
+end
+
+function g.test_http_event_removing()
```
Comment (rosik) [inline test/common_test.lua]: Why do you place it in the `common_test.lua`?

File: test/common_http_test.lua [hunk 15]
```diff
@@ -0,0 +1,63 @@
+local fio = require('fio')
+local t = require('luatest')
+local g = t.group()
```
Comment (rosik) [inline test/common_http_test.lua]: What does `common` stand for in `common_http`?

File: extensions.lua (hunk 11 above)
Comment (artur-barsegyan) [reply extensions.lua]: Let's leave it as is. Handling other corner cases is too redundant.

File: test/common_test.lua (hunk 6 above)
Comment (artur-barsegyan) [reply test/common_test.lua]: ???

File: extensions.lua (hunk 12 above)
Comment (artur-barsegyan) [reply extensions.lua]: I think that's OK. I've tried to rewrite this piece of code but it led to more complex constructions.

File: extensions.lua (hunk 11 above)
Comment (rosik) [reply extensions.lua]: Are you kidding? Imagine the cluster with HTTP enabled on routers and disabled on storage.

File: test/common_test.lua (hunk 6 above)
Comment (rosik) [reply test/common_test.lua]: It's used twice in the function. It looks reasonable to reuse it.

File: extensions.lua (hunk 11 above)
Comment (artur-barsegyan) [reply extensions.lua]: Done

**Comment by rosik** [comment]
Closing in favor of #9
